### ISSUE TITLE
No destination workload labels when using fault injection

### ISSUE THREAD

**Issue by Stono** [istio/istio#21923]
**Bug description**
We've started using istios fault injection on chaos days at AutoTrader.  However we're then asking people to investigate their dashboards in order to understanding the cause of the faults they're seeing, but none of them work as metrics reported with faults are missing workload labels.

EG:

`app-a` -> `app-b` = fine.

Add 10% 500 failure-rate to `app-b`

and `destination_workload` etc are `unknown`.
<img width="539" alt="Screenshot 2020-03-06 at 14 25 13" src="https://user-images.githubusercontent.com/1486729/76091877-4edcc800-5fb6-11ea-965c-5883d9678d60.png">

**Expected behavior**
Workload labels to work with fault injection

**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.4.6
 
**How was Istio installed?**

**Environment where bug was observed (cloud vendor, OS, etc)**

**Comment by Stono** [comment]
The other thing I've noticed as well is that `destination_service` is fully qualified (eg `app.namespace.svc .cluster.local`) in normal metrics but short formed (eg `app.namespace`) in `FI` metrics, which results in a bunch more cardinality

**Comment by Stono** [comment]
For anyone coming to this thread, this is my really hacky prometheus work around based on our opinionated 1 app-per-namespace setup.

```
    # These relabels take metrics which dont have workload information, and adds it based on the destination service
  - source_labels: [destination_service_name]
    regex: '(BlackHoleCluster)'
    replacement: '$1'
    target_label: destination_workload

  - source_labels: [destination_service, destination_app]
    regex: '^([^.]+)\.([^.]+)(\.svc\.cluster\.local)?;unknown'
    replacement: '$1'
    target_label: destination_service_name

  - source_labels: [destination_service, destination_app]
    regex: '^([^.]+)\.([^.]+)(\.svc\.cluster\.local)?;unknown'
    replacement: '$2'
    target_label: destination_service_namespace

  - source_labels: [destination_service, destination_app]
    regex: '^([^.]+)\.([^.]+)(\.svc\.cluster\.local)?;unknown'
    replacement: '$2'
    target_label: destination_workload

  - source_labels: [destination_service, destination_app]
    regex: '^([^.]+)\.([^.]+)(\.svc\.cluster\.local)?;unknown'
    replacement: '$2'
    target_label: destination_workload_namespace

  - source_labels: [destination_service, destination_app]
    regex: '^([^.]+)\.([^.]+)(\.svc\.cluster\.local)?;unknown'
    replacement: '$2'
    target_label: destination_app
```

**Comment by kyessenov** [comment]
The root cause is that fault injection happens prior to load balancing. So there's no destination endpoint to report on. We'd need to re-think how faults are injected to get a more accurate picture.

**Comment by Stono** [comment]
not stale!

**Comment by Stono** [comment]
@kyessenov @howardjohn can we reopen this please?  it's still a issue

**Comment by mandarjog** [comment]
Yes, we should have a design in 1.7 timeframe.

**Comment by kyessenov** [comment]
Two proposals:
1. Re-design fault injection as an istio extension. Consider applying it server-side by default (which makes more sense).
2. Patch fault injection xDS with some overrides in filter state or metadata that telemetry can pick up. It doesn't address the issue that fault injection extension does not play well by itself.

**Comment by mandarjog** [comment]
We should Redesign fault injection to be server side .

**Comment by howardjohn** [comment]
Why should fault injection be server side? I think it makes sense on the client, you may have different settings per client

**Comment by kyessenov** [comment]
@howardjohn Client side overrides make sense, but should not be the default. As a service owner, I shouldn't trust every client to implement my service features. The important point is to move fault injection out of the core to an extension, as it really is an envoy extension.

**Comment by douglas-reid** [comment]
As I've expressed in other forums, I expect the switch from client-side to server-side for fault injection to be somewhat controversial and, potentially, problematic.

I think targeting 1.7 for a full redesign of this feature is a bit ambitious at this point, given that we'll have to drive consensus across the entire community. This isn't on our roadmap for 1.7 and wasn't vetted by the TOC.

If we are serious about working on this for 1.7, we need to decide on a priority and get TOC sign-off on the work and priority.

**Comment by Stono** [comment]
I’m personally not too fussed on the priority as I have a hacky work around, but I think being transparent about it regression is important as this wasn’t a problem on telemetry v1.

As a user it does worry me that one features progression (telemetryv2) can lead to another features regression (faulty injection) and then it takes a re-architecture and TOC sign off, to address what is at its core, a regression/bug. 

Perhaps clarity around fault injections status on https://istio.io/latest/about/feature-stages/ combined with a “known issues” section might be a start? 🤷🏻‍♂️

**Comment by kyessenov** [comment]
@douglas-reid I really don't see how to make destination workload labels work without client attempting to reach a destination, meaning fault injection happens on the server. 

If we can scope it down to specific attributes (just namespace, or deployment), pilot could infer the common denominator in all endpoints and stick them in. That would be a temporary work around.

**Comment by douglas-reid** [comment]
@kyessenov maybe the work on out-of-mesh metadata will help here? I'm not trying to argue that we shouldn't do the work or that it is unnecessary -- only that this isn't a small fix, but rather a behavioral change that has impacts outside of telemetry, and thus requires outside input and consensus.

To the extent that we can address this with workarounds in the short-term, we absolutely should. As @Stono points out, we ideally shouldn't need to change the way FI is working just to preserve behavior.

**Comment by kyessenov** [comment]
Fair enough @douglas-reid . There is a category of telemetry gaps related to protocol errors. FI is basically causing a protocol error due to unexpected failure to exchange metadata. I think we can justify adding extra client metadata if we can phrase it as an improvement of telemetry for communication errors (client using http/1.0 or http/1.1 to talk to h2 server, websockets, client using TLS when server is not or vice versa, and so on). We would need to populate xDS metadata to create an incomplete peer metadata info. We'd need to think how FI affects cluster selection, since clusters look like the better choice to host metadata.

**Comment by Stono** [comment]
@howardjohn @douglas-reid can we reopen please, this isn't stale :s it's still an issue

**Comment by mandarjog** [comment]
Still an issue.

**Comment by bianpengyuan** [comment]
> The other thing I've noticed as well is that destination_service is fully qualified (eg app.namespace.svc .cluster.local) in normal metrics but short formed (eg app.namespace) in FI metrics, which results in a bunch more cardinality
FWIW, ^^ has been resolved in 1.8. FI metric now should have the same service labels as metrics for normal requests. Workload labels are still unknown.

**Comment by chinaran** [comment]
We have same issue. 
When using fault injection at multiple istio cluster, "destination_cluster" value is "unknown" too, so we cannot distinguish which cluster the destination service belongs to.
